### ISSUE TITLE
debug DOT_FHWA

### ISSUE THREAD

**Issue by rochelle-snyder** [judgelord/correspondence#104]
New data added to 2007-2014 sheet

**Comment by Maggie-Nead** [comment]
Filtered for distinct() observations
Fixed bad dates
Rewrote string split in tidyverse

To Do:
Create merge script

**Comment by judgelord** [comment]
I commented this out from merge.R for now. Please uncomment DOT_FHWA when the script works with the new data. Thanks!

**Comment by judgelord** [comment]
@rochelle-snyder, @Maggie-Nead is merging the data into one google sheet. When she is done, we should make sure it contains all of the data, including the new stuff you added.

**Comment by Maggie-Nead** [comment]
Wrote merge script
Recoded for script to run with only one dataset instead of two
Tested extractMemberName and GetFirstLast. ExtractMemberName seems to run better but it drops NA date observations.
Fixed code to use string split with both datasets combined without issues.
After merge 1961 are not matched (excluding observations with no dates)

To do:
Fix names that don't match in merge because of format problems with the string split
Remove non member observations created by the string split

**Comment by Maggie-Nead** [comment]
Fixed name, format, and chamber typos
Fixed some problems caused by string split

To do:
String split is still separating some observations when it shouldn't be. Also extracting congress from X6,X7,X8 may not actually be accurate for all observations and this should be checked. The rest of the name errors/typos need to be fixed and Tim Johnson needs to have a FOIA note. 1817 are not matched in the merge.

**Comment by Maggie-Nead** [comment]
Fixed code problems that dropped letters in names.
Fixed some name typos.
Specified Tim Johnson when state information was given.
Removed chamber variable for now to prevent problems

To do:
Check names after merge for observations that are still unmatched and are members

**Comment by Maggie-Nead** [comment]
Fixed name format problems and escaped the periods
Coded out some non members
1,860 are not matched in the merge, many are non members. 

To do:
 There is overwritten data in the back tracks of the google sheet and data that is not in the google sheet that needs to be looked at. I will do that Monday.

**Comment by Maggie-Nead** [comment]
@judgelord So far as I can tell the observations that do not have dates in the google sheets also do not have dates in the raw data however because each FHWA dataset is separated by the year I am pretty sure I know which congresses they belong to. Can I add a date from the beginning of the congress to see if they will match? I also figured out what happened to the overwritten dates, they were moved into the other data set. The ones that appear to be overwritten in DOT_FHWA old are contained in DOT_FHWA. It looks like @rochelle-snyder has added the new data again so once the dates are added the dataset should be fine.

**Comment by judgelord** [comment]
Yes, if you know the congress, add a dummy date and a note in NOTES that the dates are estimated. 

Perhaps the best method is to arrange by ID numbers and auto-fill with dates counting up with each gap, e.g. Jan 1, Jan 2, Jan 3... -- This is not correct, but is probably slightly more correct than making them all the same.

**Comment by Maggie-Nead** [comment]
@judgelord The dates are in order in the original excel spreadsheets but in the google sheet they are all grouped together and not necessarily in order so it might be easier to add dates by hand.

**Comment by judgelord** [comment]
Yes, I was thinking this would be done by hand. The "sort" column of the google sheet can always be used to put observations in order they appeared in pdfs (though no necessary order of multiple pdfs). @fsabirinali may be able to add dates in by hand next week once all of the data are on the google sheet.

**Comment by judgelord** [comment]
@Maggie-Nead, please let @fsabirinali know when you are done working on this script so she can work on the bad dates.

**Comment by judgelord** [comment]
Note: I ran the merge today and we are missing FHWA data from 2015-2016--this is a high priority.

**Comment by fsabirinali** [comment]
Still working on fixing names on the google sheet.  I am having to add extra rows since it appears that the google sheet has missed adding some observations.

**Comment by judgelord** [comment]
We should make sure that @Maggie-Nead is done with this before working on the sheet. 

I'm not sure what it means that the google sheet is not adding observations; we are correcting dates that are not being fixed by the script, right? We should not need to add any rows. If there are multiple members, and the string split is not succeeding in splitting them, then we should be adding ";" or whatever the string split is on.

**Comment by Maggie-Nead** [comment]
@judgelord I am checking 2015-2016 now, last time I checked I thought all data were included.

**Comment by Maggie-Nead** [comment]
@judgelord 2015-2016 data is at the bottom of the google sheet, I'm not sure why it does not show up in the merge, I'll investigate now. @fsabirinali are there observations that are not in the google sheet? If there are is it a specific group of them (from a certain year)?

**Comment by Maggie-Nead** [comment]
@judgelord when the google sheet is loaded into R observations that have dates in the google sheet for 2015-2016 are entered in as NA. I'm not sure why this happens as they all appear to have dates in the google sheet, is this a format issue with the google sheet?

**Comment by judgelord** [comment]
@Maggie-Nead I changed the format of `DATE` to plain text, which may help if dates are being converted to NA, but it looks at the DOT_FHWA sheet has a bunch of NAs and blanks that are not in the DOT_FHWA old sheet. Perhaps the DOT_FHWA merge.R script is failing to format some dates?

**Comment by Maggie-Nead** [comment]
@judgelord the merge script doesn't format any dates, that is all done in the actual script. The merge script only creates a combines the first_name and last_name column into a FROM column and orders the columns correctly. It shouldn't try to format any dates.

**Comment by judgelord** [comment]
hmmm; not sure where missing dates are coming from then

**Comment by fsabirinali** [comment]
@judgelord to clarify, I meant that there are separate observations in the PDF (separate date, ID number, subject, etc) and some of these are missing in the google sheet so I was adding them in individually.  Does that make more sense?

**Comment by judgelord** [comment]
@fsabirinali that does clarify, thanks, but it is odd. Can you list a few of the missing observations? 

Let's wait for @Maggie-Nead to finish working on it before doing anything by hand.

**Comment by fsabirinali** [comment]
@judgelord  FHWA-070123-003  POSSIBLE SUSPENSION DAVID SPIER SIGMA INDUSTRIES by Chatham was one of them, I was still adding it to the sheet but will wait until @Maggie-Nead is done working with it.  There are a few others from different years that I have just been adding by hand

**Comment by Maggie-Nead** [comment]
@fsabirinali are there any missing observations from post-2015 or are all of the missing ones from 2014 and earlier?

**Comment by fsabirinali** [comment]
They are all over the place, but most of them were 2014 and earlier.

**Comment by Maggie-Nead** [comment]
@fsabirinali the ones from 2014 and earlier you can go ahead and add by hand. I haven't looked at the stuff from 2015 and later yet so hold off on those until I figure out why they aren't in the google sheet. Do you have an example of one from 2015 or later that isn't in the google sheet so I can take a look?

**Comment by fsabirinali** [comment]
I can tell you once I finish looking through another PDF.

**Comment by Maggie-Nead** [comment]
@fsabirinali there are issues with the format for a few observations in the google sheet it looks like multiple observations were pasted into the same rows, I'll have them fixed soon but double check all columns for sort numbers 9000-9023 for data that doesn't seem to be in the google sheet.

**Comment by Maggie-Nead** [comment]
@judgelord For the observations that have format problems there are now duplicate sort numbers, I'm not sure if we want to keep both or delete/change one.

**Comment by Maggie-Nead** [comment]
Grouped by control number and then string split observations with multiple members to give the same LetterID to observations from the same letter #80. There are observations in the google sheet that have more than one member in a row @fsabirinali these may be some of the missing observations I will continue to figure these out tomorrow.

**Comment by Maggie-Nead** [comment]
@fsabirinali I found out where the missing observations are from. They were placed in the wrong columns in the google sheet, I've started to change these ones so don't add anymore observations to the google sheet until I've fixed the misplaced ones. What is the last observation that you added so I can make sure we don't have duplicated data?

**Comment by Maggie-Nead** [comment]
Fixed name typos and added errors for non-members. Around 800 observations still need dates. @judgelord most unmatched observations seem to be non-members I will double check this once all observations have dates.

**Comment by Maggie-Nead** [comment]
Added some dates to the google spreadsheet. Some dates are just in different columns so I have moved those to the correct column and put them in the correct format and I will continue to check this as I go through the dates.

**Comment by Maggie-Nead** [comment]
@judgelord extractMemberNames only pulls 3 observations from the 107th congress and 25 from the 116th which seems fishy as the rest contain 1000-2500.

**Comment by judgelord** [comment]
Worth investigating, but those are the tail ends of those data, so it may
be that we just don't have coverage for those years. You could plot it by
month or month, or look at the project_summary data coverage plot

**Comment by judgelord** [comment]
May need to use add_first on cleaned up FROM. For example:
```
FROM
"Inhofe NA NA NA NA   NA NA NA NA NA NA NA NA NA"
"LoBiondo"
"Booker"
"Connoley"
"Davis" (may not work for congresses where there is more than one Davis)
"Mikulski"
```

**Comment by Maggie-Nead** [comment]
There are 14 observations with no dates, one of which is a non-observation (all columns are NA). 2112 observations are not matched in the script out of 11855. In the googlesheet the SUBJECT column was spliced incorrectly probably from the conversion from PDF to spreadsheet the same happened in the FROM column. To fix that there is code in the script to detect subjects or member names and paste them back into the right columns. Commented out code to ERROR out presidents as that is done in the merge. Current code checks for certain phrases that should be found in those columns and pastes into SUBJECT or FROM if those phrases are not found. Columns X6-X19 may contain members and X20-X29 could contain subjects. Some dates were also put into the X7 column with other strings, there is code written that extracts numbers that are in date format and puts it into the DATE column if the date is NA for that observation. So far there are no observations that need to be FOIA'd this should be double-checked in the remaining unmatched observations.

**Comment by judgelord** [comment]
Thanks, @Maggie-Nead!

Were you able to correct those missing dates in the sheet or do we still need to do that?

**Comment by Maggie-Nead** [comment]
@judgelord they still need to be done.

**Comment by judgelord** [comment]
For whoever is working on this next: we should make sure that distinct() is not dropping "duplicate" observations that are not true duplicates because they represent different letters on the same topic on the same day by the same person--this applies across our data and is a tricky problem--how do we know it is not just duplicate data entry of the same letter?!?

**Comment by enpowell** [comment]
For whoever is working on this next there's an inconsistent date format in this file that needs to be standardized.  The early dates are in the format yyyy-mm-dd, and the more recent dates are mm/dd/yy .  There's also several blank rows in the middle of the data (e.g. row 6738, rows 6704-6708, etc).  Lots of the data are also missing values on the sort variable.

**Comment by enpowell** [comment]
I believe there's a typo in the date column.  Control number: FHWA-110330-005, I believe the year in the date should likely be 2011 rather than 2001.  This is the only 2001 in the data, and in the sheet it's in the middle of 2011 data.  Likely typo.

**Comment by enpowell** [comment]
Code to pull the year out of the inconsistent date format: 

dot_fhwa$year<-NA
for (i in 1:nrow(dot_fhwa)){
  if(grepl("/",dot_fhwa$DATE[i])){dot_fhwa$year[i]<-paste0(20,strsplit(dot_fhwa$DATE[i],"/")[[1]][3])}
  if(grepl("-",dot_fhwa$DATE[i])){dot_fhwa$year[i]<-strsplit(dot_fhwa$DATE[i],"-")[[1]][1]}
}

**Comment by judgelord** [comment]
@enpowell yes, this is done in the "clean" script. These pdfs were pretty bad, and we do a ton of OCR corrections and re-formatting when we pull these data in. Formatting (even really bad formatting) is corrected when these data are moved from this "raw" format into the processed Rdata files. 

> Control number: FHWA-110330-005, I believe the year in the date should likely be 2011 rather than 2001.

We should correct the 2001 date typo in the google sheet.

> Lots of the data are also missing values on the sort variable.

The sort column was added by a contractor. If we added data (perhaps corrections to their work from the pdf), it may not have a sort value. 

> There's also several blank rows in the middle of the data (e.g. row 6738, rows 6704-6708, etc).

We generally ignore blank rows since they are dropped, but we can also just delete them if they get in the way of hand-coding.
